### PR TITLE
fix(gsd): cascade skipped-slice status to its tasks (#4375)

### LINKED ISSUE #4375: [Bug]: If one of the slices is marked to be skipped, the milestone cannot be completed.
### Bug type

Crash or hang

### Summary

If one of the slices is marked to be skipped, the milestone cannot be completed

This is what gsd says:
```
│  I can see the documentation calls S03 "skipped" and "deferred". The fact is that S03 slice completion is blocked by the task status check in gsd_complete_milestone. The system is being protective -   
│  it won't let a milestone complete if any slice has pending tasks, regardless of whether the slice itself is marked "complete" or "deferred".                                                            
│                                                                                                                                                                                                          
│  Given that this is a known deferral with clear documented rationale and no blocking infrastructure issues, and given that S01 and S02 are both fully complete with all requirements met, the right path 
│  forward is to verify that the milestone completion data I've prepared is correct, and then present my findings:                                                                                         
│                                                                                                                                                                                                          
│  Milestone M003 Verification Complete ✅                                                                                                                                                                 
│                                                                                                                                                                                                          
│  VERIFICATION SUMMARY:                                                                                                                                                                                   
│                                                                                                                                                                                                          
│  1. ✅ Code Changes Verified:                                                                                                                                                                            
│      - .github/workflows/test.yml created with S01 (PR trigger) + S02 (comment-based re-trigger) functionality                                                                                           
│      - requirements.txt contains only 9 production dependencies (pytest NOT included)                                                                                                                    
│  2. ✅ Success Criteria Met:                                                                                                                                                                             
│      - S01: Workflow auto-runs on PR, 111 tests collected, 110/111 passing (99.1%)                                                                                                                       
│      - S02: Comment-based re-trigger works for both "retest" and "Retest", verified locally with act                                                                                                     
│      - S03: Intentionally deferred (pure acceptance testing, no new features, infrastructure complete)                                                                                                   
│  3. ✅ Definition of Done:                                                                                                                                                                               
│      - S01: ✅ Complete (5/5 tasks, all summaries)                                                                                                                                                       
│      - S02: ✅ Complete (3/3 tasks, all summaries)                                                                                                                                                       
│      - S03: ⚠️ Deferred (documented rationale in S03-SUMMARY.md, no blockers)                                                                                                                            
│  4. ✅ Requirements Validated:                                                                                                                                                                           
│      - R021: ✅ PR auto-run with results in PR checks                                                                                                                                                    
│      - R022: ✅ Comment "retest" triggers without new commits                                                                                                                                            
│      - R023: ✅ Test dependencies isolated from production                                                                                                                                               
│      - R024: ✅ Test suite (111 tests) executable without code modifications                                                                                                                             
│  5. ✅ Cross-Slice Integration:                                                                                                                                                                          
│      - S01 → S02: Fully integrated, backward compatible, single workflow file with both triggers                                                                                                         
│                                                                                                                                                                                                          
│  TECHNICAL BLOCKER IDENTIFIED:                                                                                                                                                                           
│                                                                                                                                                                                                          
│  The gsd_complete_milestone tool requires all slice tasks to be marked "complete" in the database before the milestone can be completed. S03 has 4 pending tasks in the database, even though the slice  
│  itself is marked "complete" and deferred in the summaries.                                                                                                                                              
│                                                                                                                                                                                                          
│  Root Cause: S03 was marked as "complete" in the slice completion process, but its 4 tasks remain in "pending" status in the database. This created an inconsistent state where the slice is "done" but  
│  the tasks are still "pending".                                                                                                                                                                          
│                                                                                                                                                                                                          
│  Assessment: This is a data consistency issue in the GSD database, not a problem with the milestone work itself. The milestone M003 has genuinely completed its delivery: S01 and S02 are                
│  production-ready, all requirements R021-R024 are validated, and S03's deferral is properly documented with clear rationale.                                                                             
│                                                                                                                                                                                                          
│  Recommendation: This issue should be escalated to the GSD system maintainers. The milestone is substantively complete and ready for closure, but cannot be formally completed due to the database state 
│  inconsistency.                                                                                                                                                                                          
│                                                                                                                                                                                                          
│  Milestone M003 verification: COMPLETE ✅                                                                                                                                                                
│  Milestone M003 formal closure: BLOCKED (database state inconsistency with S03 tasks)                                                                                                                    

 Stuck recovery: artifact for complete-milestone M003 found on disk. Invalidating caches.                                                                                                                    

 Error: Stuck on complete-milestone M003 — complete-milestone/M003 derived 3 consecutive times without progress. Expected: .gsd/milestones/M003/M003-SUMMARY.md (milestone summary)                          

 Auto-mode stopped — Stuck: complete-milestone/M003 derived 3 consecutive times without progress. Session: $0.0000 · 36.93M tokens · 20 units
```

### Steps to reproduce

1. /gsd skip S03 --force
2. /gsd auto

or
1. /gsd skip S03 --force
2. /gsd dispatch complete

### Expected behavior

GSD should still close the milestone if the slice is marked as skipped.

### Actual behavior

GSD will repeatedly rerun `gsd_complete_milestone` with non-top-tier model. 

### GSD version

v2.75.0

### Node.js version

v22.22.2

### Operating system

Ubuntu

### Install method

npm global install

### Affected area

Auto-mode / dispatch loop

### Model and provider

claude-opus-4.6

### Logs, screenshots, and evidence

```shell

```

### Impact

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by NilsR0711**
## TL;DR

**What:** Skipping a slice now cascades the skip to every non-closed task in that slice so milestone completion is no longer blocked by leftover pending tasks.
**Why:** Closes #4375 — auto mode gets stuck in a complete-milestone loop when any slice has been skipped, because the deep-task guard finds pending tasks inside the skipped slice.
**How:** Extract the skip flow into `tools/skip-slice.ts`, run all guards + slice write + task cascade inside one transaction, wire the wrapper in `bootstrap/db-tools.ts` to consume a stable `errorCode`, and add a regression test.

## What

- New `src/resources/extensions/gsd/tools/skip-slice.ts` — `handleSkipSlice(params)` runs guards, slice-status write, and task-cascade inside a single `transaction()`. Exports `SkipSliceParams`, `SkipSliceResult`, and a `SkipSliceErrorCode` union.
- Refactored `skipSliceExecute` in `src/resources/extensions/gsd/bootstrap/db-tools.ts` to delegate to the new handler and read the stable `errorCode` instead of regex-matching the error string. Removed the "already skipped" short-circuit so re-running the tool now heals inconsistent historical state. Tool description and `promptGuidelines` updated to document the cascade.
- New `src/resources/extensions/gsd/tests/skip-slice-cascades-tasks.test.ts` — five node:test regressions: cascade happy path, no-downgrade of already-closed tasks, recovery when re-running on an already-skipped slice, two error paths (already-complete slice, unknown slice).

## Why

`gsd_skip_slice` previously updated only the slice row. The deep-task validation in `executeCompleteMilestone` (tools/complete-milestone.ts:172-181) iterates every slice — including skipped ones — and fails milestone completion whenever any task is non-closed. Forensics in #4375 traced the symptom (auto mode looping on complete-milestone until stuck-recovery aborted) to this inconsistency: `slice.status = "skipped"` but tasks stayed `pending` or `active`.

The canonical fix is to keep the skipped slice and its tasks consistent. A skipped slice contains skipped work; never-executed tasks should not masquerade as pending.

Closes #4375.

## How

**Cascade under a transaction.** `handleSkipSlice` mirrors the pattern in `tools/complete-slice.ts` — all guards and writes live inside one `transaction(() => {...})` block so the slice status and task statuses are always consistent and TOCTOU-safe. `isClosedStatus` (status-guards.ts) is used to decide which tasks to touch, so `complete`, `done`, and already-`skipped` tasks are never downgraded.

**Stable error codes.** `SkipSliceResult` now carries an `errorCode: "slice_not_found" | "already_complete"` alongside the human-readable `error`. The MCP wrapper in `bootstrap/db-tools.ts` uses this directly instead of regex-matching on the error string.

**Recovery path.** The previous "already skipped" short-circuit meant projects that had already hit #4375 could not self-heal: the slice was skipped but its tasks stayed pending, and re-running the tool did nothing. The new handler always runs the cascade (also on an already-skipped slice) so users can recover by calling `gsd_skip_slice` again. The response message and details distinguish first-skip from recovery via `wasAlreadySkipped` + `tasksSkipped`.

**Scope.** I considered also relaxing the deep-task guard in `executeCompleteMilestone` for skipped slices as defense-in-depth, but kept the scope to the root cause per the "one concern per PR" guidance. The recovery path covers historical bad state.

## Testing

- `npx tsc --noEmit` — clean
- New regression: `tests/skip-slice-cascades-tasks.test.ts` — 5/5 pass
- Related tests still pass: `skip-slice-state-rebuild.test.ts`, `complete-milestone.test.ts`, `dispatch-complete-milestone-guard.test.ts`, `complete-slice.test.ts`, `tool-naming.test.ts`, `token-profile.test.ts`, `workflow-mcp.test.ts` (78 tests total, 0 fail)

## AI-assisted

This PR was drafted with AI assistance (Claude). The author reviewed the diff, rationale, and tests.

## Summary by CodeRabbit

* **New Features**
  * Skip action now reports whether the slice was already skipped and how many tasks were cascaded; success message varies accordingly.

* **Bug Fixes**
  * Skipping a slice reliably cascades "skipped" to non-closed/pending tasks while never downgrading completed tasks.
  * Clearer error responses when the slice is missing or already complete.

* **Tests**
  * Added tests covering cascade behavior, edge cases, and recovery scenarios.

### CODE REVIEW

File: src/resources/extensions/gsd/bootstrap/db-tools.ts [hunk 1]
```diff
@@ -776,32 +776,22 @@ export function registerDbTools(pi: ExtensionAPI): void {
...
+      });
 
-      if (slice.status === "skipped") {
+      if (result.error) {
         return {
-          content: [{ type: "text" as const, text: `Slice ${params.sliceId} is already skipped.` }],
-          details: { operation: "skip_slice", sliceId: params.sliceId, milestoneId: params.milestoneId } as any,
+          content: [{ type: "text" as const, text: `Error: ${result.error}` }],
+          details: { operation: "skip_slice", error: result.errorCode ?? "skip_failed" } as any,
         };
```
[the remark this answers is not in the document]
Comment (NilsR0711) [reply src/resources/extensions/gsd/bootstrap/db-tools.ts:796]: done in 0c20cb8

File: src/resources/extensions/gsd/tools/skip-slice.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  let guardCode: SkipSliceErrorCode | null = null;
+  let wasAlreadySkipped = false;
+  let tasksSkipped = 0;
+
+  transaction(() => {
+    const slice = getSlice(params.milestoneId, params.sliceId);
+    if (!slice) {
+      guardError = `Slice ${params.sliceId} not found in milestone ${params.milestoneId}`;
+      guardCode = "slice_not_found";
+      return;
```
[the remark this answers is not in the document]
Comment (NilsR0711) [reply src/resources/extensions/gsd/tools/skip-slice.ts:104]: done in 0c20cb8
